### PR TITLE
allow comment in multiple line pipeline

### PR DESCRIPTION
**Pull request by WindSoilder**
# Description
- fixes: #5517
- fixes: #9250

For the following commands:
```
ls
# | le
| length
```

I found that it generates a bad lite parsing result:
```
LiteBlock { 
    block: [
        LitePipeline { 
            commands: [
                Command(None, LiteCommand { comments: [], parts: [Span { start: 138600, end: 138602 }] })
            ]
        },
        LitePipeline { 
            commands: [
                Command(Some(Span { start: 138610, end: 138611 }),
                LiteCommand { comments: [Span { start: 138603, end: 138609 }], parts: [Span { start: 138612, end: 138618 }] })
            ]
        }
    ]
}
```

Which should contains only one `LitePipeline`, and the second `LitePipeline` is generated because of `Eol` lex token:
```
[
    Token { contents: Item, span: Span { start: 138600, end: 138602 } },
    Token { contents: Eol, span: Span { start: 138602, end: 138603 } },    // it generates the second LitePipeline
    Token { contents: Comment, span: Span { start: 138603, end: 138609 } },
    Token { contents: Pipe, span: Span { start: 138610, end: 138611 } },
    Token { contents: Item, span: Span { start: 138612, end: 138618 } }
]
```

To fix the issue, I remove the `Eol` token when we meet `Comment` right after `Eol`, then it will generate a good LiteBlock, and everything will work fine.

### After the fix:
Token:
```
[
  Token { contents: Item, span: Span { start: 138618, end: 138620 } },
  Token { contents: Comment, span: Span { start: 138622, end: 138628 } },
  Token { contents: Pipe, span: Span { start: 138629, end: 138630 } },
  Token { contents: Item, span: Span { start: 138631, end: 138637 } }
]
```

LiteBlock:
```
LiteBlock {
  block: [
    LitePipeline { 
      commands: [
        Command(
            None, 
            LiteCommand { 
                comments: [Span { start: 138622, end: 138628 }],
                parts: [Span { start: 138618, end: 138620 }] 
            }
        ),
        Command(
            Some(Span { start: 138629, end: 138630 }),
            LiteCommand { comments: [], parts: [Span { start: 138631, end: 138637 }] })] }] }
```

# User-Facing Changes

# Tests + Formatting

# After Submitting

### CODE REVIEW

**Comment by hustcer** [comment]
Nice, I have been waiting for this feature for a long time

**Comment by WindSoilder** [comment]
Oh, it breaks module documentation..

But yeah, I really love our testing code :-D

File: crates/nu-parser/src/lex.rs [hunk 1]
```diff
@@ -438,6 +438,19 @@ fn lex_internal(
...
             let mut start = curr_offset;
 
+            // if previous token is eol, we just delete that token
+            // so at high level, it makes the following transform:
+            //
+            // ls             ls  # | le
+            // # | le    ==>  | length
+            // | length
+            //
+            // And it makes lite_parsing easiler to generate a correct LiteBlock
```
Comment (amtoine) [inline crates/nu-parser/src/lex.rs:448]: should be *easier* i think :yum:

Comment (amtoine) [inline crates/nu-parser/src/lex.rs:448]: maybe we can just link to this PR instead of adding a huge comment? :thinking:

**Review by amtoine — APPROVED** [review]
oooooh wow, i love this fix, been waiting for it a long time too :scream: :heart: 

i've tried the following snippet, as in the code example
```bash
ls
# | le
| length
```

i've tried in the following cases and it only works in the REPL :open_mouth: 
- :heavy_check_mark: REPL
- :x: `nu script.nu`
- :x: `nu -c "..."`

**Comment by WindSoilder** [comment]
Mark as draft for now because it seems not good for existing module document

File: crates/nu-parser/src/lex.rs (hunk 1 above)
Comment (WindSoilder) [reply crates/nu-parser/src/lex.rs]: Hi, I've reimplement it in another place :)

Comment (WindSoilder) [reply crates/nu-parser/src/lex.rs]: Normally I'm not a fans of link a pr as a comment, it's poor explained if the it's not too complex

File: crates/nu-parser/src/lex.rs [hunk 2]
```diff
@@ -388,7 +388,24 @@ fn lex_internal(
...
+                        //
+                        // Initially, the last one token is TokenContents::Pipe, we don't need to
+                        // check it, so the beginning offset is 2.
+                        let mut offset = 2;
+                        while output.len() > offset {
+                            let index = output.len() - offset;
+                            if output[index].contents == TokenContents::Comment
+                                && output[index - 1].contents == TokenContents::Eol
+                            {
+                                output.remove(index - 1);
```
Comment (sholderbach) [inline crates/nu-parser/src/lex.rs:403]: Quick question as I am not familiar with our lexer output: 

Is our lex output intended to be contiguous (i.e. all bytes covered by tokens) or do we only report parts with parseable tokens and skip whitespace? 
Just to make sure we don't run afoul in any span dereferencing.

**Review by sholderbach — APPROVED** [review]
Thanks so much! That looks like it removes a major limitation!

File: crates/nu-parser/src/lex.rs (hunk 2 above)
Comment (sholderbach) [reply crates/nu-parser/src/lex.rs:403]: Great! Thanks!

**Comment by amtoine** [comment]
thanks a million @WindSoilder, this is truely amazing and so useful to test things easily without breaking pipelines :star_struck: 

i've tested
```bash
ls
# | le
| length
```
in the REPL, in a script and in `nu -c "..."`
and also
```bash
export def main [] {
    ls
    # | le
    | length
}
```
as a module.

in all these cases, i consistently get the same answer as `ls | length` in the REPL... noishe :triumph:

**Pull request merged**